slurm_openstack_tools/reboot.py

```diff
@@ -1,160 +1,122 @@
-        )
-
-    image = conn.image.find_image(image_descr)  # doesn't throw exception
-    if image is None:
-        logger.error(f"image {image_descr} either not found or not unique")
+def get_hostlist_from_slurm():
+    """
+    Parse the hostlist provided by Slurm as command-line arguments.
+    """
+    if len(sys.argv) < 2:
+        logger.error("No hostlist provided by Slurm")
         sys.exit(1)
-
-    # Note that OpenStack will power down the server as part of the rebuild
-    logger.info(f"rebuilding server %{server_id} with image %{image.id}")
-    conn.rebuild_server(server_id, image.id)
+    return sys.argv[1].split(",")
 
 
-def do_reboot():
-    # in case we are on an NFS share,
-    # try to ensure python stopped before reboot
-    os.execvp("reboot", ["reboot"])
+def read_hostvars(node):
+    """
+    Read the hostvars.yml file for a specific node.
+    """
+    hostvars_file = Path(HOSTVARS_DIR) / node / "hostvars.yml"
+    if not hostvars_file.exists():
+        logger.warning(f"No hostvars.yml found for node: {node}")
+        return None
 
+    with open(hostvars_file, "r") as f:
+        return yaml.safe_load(f)
 
-def rebuild_or_reboot():
-    """A RebootProgram for slurm which can rebuild the node running it.
 
-    This is intended to set as the `RebootProgram` in `slurm.conf`.
-    It is then triggered by slurm using something like:
-        scontrol reboot [ASAP] reason="rebuild image:<image_name_or_id>" <NODES>
+def get_current_image(conn, server_id):
+    """
+    Retrieve the current image for a server using OpenStack.
+    """
+    server = conn.get_server(server_id)
+    if not server:
+        logger.error(f"Server ID {server_id} not found in OpenStack")
+        return None
+    return server.image.id
+
+
+def rebuild_node(conn, server_id, target_image_id):
+    """
+    Rebuild the node with the target image.
+    """
+    image = conn.image.find_image(target_image_id)
+    if not image:
+        logger.error(f"Target image {target_image_id} not found in OpenStack")
+        return False
+
+    logger.info(f"Rebuilding server {server_id} with image {target_image_id}")
+    conn.rebuild_server(server_id, target_image_id)
+    return True
+
+
+def reboot_node(conn, server_id, reboot_type="HARD"):
+    """
+    Reboot the node (default: hard reboot).
+    """
+    logger.info(f"Rebooting server {server_id} with {reboot_type.lower()} reboot")
+    conn.compute.reboot_server(server_id, reboot_type=reboot_type)
+    return True
+
+
+def process_node(conn, node):
+    """
+    Process a single node by comparing its target and current images.
+    """
+    hostvars = read_hostvars(node)
+    if not hostvars:
+        logger.info(f"No hostvars defined for node {node}, skipping...")
+        return
+
+    server_id = hostvars.get("instance_id")
+    target_image_id = hostvars.get("image_id")
+
+    if not server_id:
+        logger.info(f"Node {node} is not an OpenStack node, ignoring...")
```

Comment:
This might want to return an exit code, so we error out, rather than silently fail. We probably need to make this know about all node types somehow, but I think failing is safer here.